src/library/scala/collection/Map.scala

```diff
@@ -29,25 +29,14 @@ trait Map[K, +V]
 
   def canEqual(that: Any): Boolean = true
 
-  override def equals(o: Any): Boolean = o match {
-    case that: Map[K, _] =>
-      (this eq that) ||
-      (that canEqual this) &&
-      (this.size == that.size) && {
-        try {
-          val checker = new AbstractFunction1[(K, V),Boolean] {
-            override def apply(kv: (K,V)): Boolean = {
-              that.getOrElse(kv._1, Map.DefaultSentinel) == kv._2
-            }
-          }
-          this forall checker
-        } catch {
-          case _: ClassCastException => false
-        }
-      }
-    case _ =>
-      false
-  }
+  override def equals(o: Any): Boolean =
+    (this eq o.asInstanceOf[AnyRef]) || (o match {
+      case map: Map[K, _] if map.canEqual(this) =>
+        (this.size == map.size) &&
+          this.forall(kv => map.getOrElse(kv._1, Map.DefaultSentinel()) == kv._2)
```

Comment:
The reification of the predicate function into `checker` was done in 035ce5843cd4817536646905a791090921e8d895, shouldn’t we keep it?

Reply:
I don't see what the remaining benefit is. [In 2.12](https://github.com/scala/scala/blob/53f8d5ec0429933099938fde1f23de6c19975236/src/library/scala/collection/GenMapLike.scala#L125) the `checker` is at the same time the predicate (`Function1`) as well as the sentinel (`Function0`). But we can avoid the lambda allocation for the default value (which is a by-name parameter) also by turning `Map.DefaultSentinel` into a `Function0`.